# Exercise 01: Statistical Bias Detection
# Cluster 3 - Statistics and Probability
# Framework: NIST AI RMF Measure function
# Case Study: Recruitment AI

In [7]:
import numpy as np

# Simulated hiring data: 100 applicants
np.random.seed(42)

# Women applicants — slightly higher average score
women_scores = np.random.normal(loc=78, scale=8, size=50)

# Men applicants
men_scores = np.random.normal(loc=76, scale=8, size=50)

print("Women scores (first 5):", women_scores[:5].round(2))
print("Men scores (first 5):", men_scores[:5].round(2))

Women scores (first 5): [81.97 76.89 83.18 90.18 76.13]
Men scores (first 5): [78.59 72.92 70.58 80.89 84.25]


In [15]:
print("Mean= for women",np.mean(women_scores))
print("standard deviation for women=",np.std(women_scores))
print("Mean for men =",np.mean(men_scores))
print("standard deviation for men=",np.std(men_scores))


Mean= for women 76.19620875795087
standard deviation for women= 7.394279502979395
Mean for men = 76.14224696374362
standard deviation for men= 6.9243006291921025


In [17]:
from scipy import stats

# T-test: compare women vs men scores
t_statistic, p_value = stats.ttest_ind(women_scores, men_scores)

print(f"T-statistic: {t_statistic:.4f}")
print(f"P-value: {p_value:.4f}")
print()
if p_value < 0.05:
    print("Result: STATISTICALLY SIGNIFICANT difference detected")
    print("Conclusion: Bias is likely REAL, not random variation")
else:
    print("Result: NO statistically significant difference")
    print("Conclusion: Difference is likely random variation")

T-statistic: 0.0373
P-value: 0.9703

Result: NO statistically significant difference
Conclusion: Difference is likely random variation


- p-value > 0.05 means no statistically significant bias in scoring
- BUT: fair scoring does not guarantee fair hiring outcomes
- NIST RMF Measure: must test scores AND outcomes AND cutoff thresholds

In [18]:
# Now simulate a BIASED system
# Women are scored lower artificially
np.random.seed(42)

women_biased = np.random.normal(loc=70, scale=8, size=50)
men_biased = np.random.normal(loc=78, scale=8, size=50)

# Run t-test
t_stat2, p_value2 = stats.ttest_ind(women_biased, men_biased)

print(f"Women mean: {np.mean(women_biased):.2f}")
print(f"Men mean: {np.mean(men_biased):.2f}")
print(f"P-value: {p_value2:.4f}")

if p_value2 < 0.05:
    print("Result: STATISTICALLY SIGNIFICANT bias detected")
else:
    print("Result: No significant difference")

Women mean: 68.20
Men mean: 78.14
P-value: 0.0000
Result: STATISTICALLY SIGNIFICANT bias detected



**Exercise 01: Statistical Bias Detection — Governance Reflection**

Statistical significance (p-value < 0.05) provides objective evidence that observed differences between groups are real and not due to random variation. In hiring systems, this matters because a regulator can demand: "Show me the data. Prove bias exists or doesn't exist." Our t-test demonstrated this — when women and men scored equally (p=0.97), no bias was detected. When women scored 10 points lower (p=0.0000), bias was statistically proven.

However, fair scoring does not guarantee fair hiring outcomes. A system can score men and women identically but still produce biased hiring if human interviewers are prejudiced, if the scoring attributes themselves encode bias (like weighting postcode as a proxy for socioeconomic status), or if the final hiring decision uses different cutoff thresholds for different groups. This is why NIST AI RMF requires four functions: we must Govern (set policies), Map (understand the system), Measure (test for bias statistically — what we did here), and Manage (monitor actual outcomes). For Recruitment AI under EU AI Act Annex III High-Risk classification, this means statistical testing of scoring alone is insufficient. Auditors must track the entire pipeline — from scoring through final hiring decisions — to prove fairness end-to-end.
